tachyon/zk/air/plonky3/base/two_adic_multiplicative_coset.h

```diff
@@ -0,0 +1,154 @@
+      return ret;
+    });
+  }
+
+  template <typename ExtField>
+  air::plonky3::LagrangeSelectors<ExtField> GetSelectorsAtPoint(
+      const ExtField& point) const {
+    ExtField unshifted_point = point * domain_->offset_inv();
+    ExtField z_h = unshifted_point.ExpPowOfTwo(domain_->log_size_of_group()) -
+                   ExtField::One();
+    std::optional<ExtField> first_row =
+        z_h / (unshifted_point - ExtField::One());
+    CHECK(first_row);
+    ExtField transition = unshifted_point - ExtField(domain_->group_gen_inv());
+    std::optional<ExtField> last_row = z_h / transition;
+    CHECK(last_row);
+    ExtField inv_zeroifier = unwrap(z_h.Inverse());
+    return {std::move(*first_row), std::move(*last_row), std::move(transition),
+            std::move(inv_zeroifier)};
+  }
+
+  air::plonky3::LagrangeSelectors<std::vector<F>> GetSelectorsOnCoset(
+      const TwoAdicMultiplicativeCoset& coset) const {
+    const F& coset_shift = coset.domain()->offset();
+
+    CHECK_EQ(domain_->offset(), F::One());
+    CHECK_NE(coset_shift, F::One());
+    CHECK_GE(domain_->log_size_of_group(), coset.domain()->log_size_of_group());
+    size_t rate_bits =
+        coset.domain()->log_size_of_group() - domain_->log_size_of_group();
+    F s_pow_n = coset_shift.ExpPowOfTwo(domain_->log_size_of_group());
+
+    // Evals of Z_H(X) = Xⁿ - 1
+    std::vector<F> evals =
+        F::GetSuccessivePowers(1 << rate_bits, domain_->group_gen(), s_pow_n);
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < evals.size(); ++i) {
+      evals[i] -= F::One();
+    }
+
+    std::vector<F> xs = F::GetSuccessivePowers(
+        coset.domain()->size(), coset.domain()->group_gen(), coset_shift);
+
+    F coset_i = domain_->group_gen().Pow(domain_->size() - 1);
+    std::vector<F> inv_denoms_inv_zeroifier = evals;
+
+    std::vector<F> first_row = xs;
+    std::vector<F> last_row = xs;
+    std::vector<F> transition = xs;
+    std::vector<F> inv_zeroifier(xs.size());
+    CHECK(F::BatchInverseInPlace(inv_denoms_inv_zeroifier));
+
+    base::Parallelize(
+        xs, [this, &evals, &coset_i, &inv_denoms_inv_zeroifier, &first_row,
+             &last_row, &transition, &inv_zeroifier](
+                absl::Span<F> chunk, size_t chunk_offset, size_t chunk_size) {
+          size_t start = chunk_offset * chunk_size;
+          std::vector<F> first_row_chunk(chunk_size);
+          std::vector<F> last_row_chunk(chunk_size);
+          OPENMP_PARALLEL_FOR(size_t i = 0; i < chunk_size; ++i) {
+            size_t idx = start + i;
+            first_row_chunk[start + i] = chunk[i] - F::One();
+            last_row_chunk[start + i] = chunk[i] - coset_i;
+          }
+          CHECK(F::BatchInverseInPlaceSerial(first_row_chunk));
+          CHECK(F::BatchInverseInPlaceSerial(last_row_chunk));
+
+          OPENMP_PARALLEL_FOR(size_t i = 0; i < chunk_size; ++i) {
+            size_t idx = start + i;
+            size_t evals_i = (start + i) % evals.size();
```

Comment:
I suggest modifying it like this:

```suggestion
            first_row_chunk[idx] = chunk[i] - F::One();
            last_row_chunk[idx] = chunk[i] - coset_i;
          }
          CHECK(F::BatchInverseInPlaceSerial(first_row_chunk));
          CHECK(F::BatchInverseInPlaceSerial(last_row_chunk));

          OPENMP_PARALLEL_FOR(size_t i = 0; i < chunk_size; ++i) {
            size_t idx = start + i;
            size_t evals_i = (idx) % evals.size();
```